src/aleo_schnorr/mod.rs

```diff
@@ -0,0 +1,137 @@
+use std::str::FromStr;
+
+use crate::crypto_tools::message_digest::MessageDigest;
+use crate::sdk::api::TofnFatal;
+use rand::SeedableRng as _;
+use snarkos_account::Account;
+use snarkvm::prelude::{Field, Network, PrivateKey, Signature, ToFields};
+
+use crate::{
+    constants::ALEO_SCHNORRR_TAG,
+    crypto_tools::rng,
+    sdk::{
+        api::{BytesVec, TofnResult},
+        key::SecretRecoveryKey,
+    },
+};
+
+#[derive(Debug)]
+pub struct KeyPair<N: Network> {
+    aleo_account: Account<N>,
+}
+
+impl<N: Network> KeyPair<N> {
+    /// tofnd needs to store this in the kv store.
+    pub fn signing_key(&self) -> &PrivateKey<N> {
+        &self.aleo_account.private_key()
+    }
+
+    pub fn encoded_verifying_key(&self) -> String {
+        self.aleo_account.address().to_string()
+    }
+}
+
+pub fn keygen<N: Network>(
+    secret_recovery_key: &SecretRecoveryKey,
+    session_nonce: &[u8],
+) -> TofnResult<KeyPair<N>> {
+    /// Domain separation for seeding the RNG
+    const KEYGEN_TAG: u8 = 0x00;
+
+    let mut rng = rng::rng_seed_signing_key(
+        ALEO_SCHNORRR_TAG,
+        KEYGEN_TAG,
+        secret_recovery_key,
+        session_nonce,
+    )?;
+
+    let private_key = PrivateKey::new(&mut rng).map_err(|_| TofnFatal)?;
+    let aleo_account = Account::try_from(private_key).map_err(|_| TofnFatal)?;
+
+    Ok(KeyPair { aleo_account })
+}
+
+pub fn sign<N: Network>(
+    signing_key: &KeyPair<N>,
+    msg_to_sign: &MessageDigest,
+) -> TofnResult<BytesVec> {
+    let message = aleo_encoded(msg_to_sign).map_err(|_| TofnFatal)?;
+
+    let sign = signing_key
+        .aleo_account
+        .sign(&message, &mut rand_chacha::ChaChaRng::from_entropy())
+        .map_err(|_| TofnFatal)?;
+
+    Ok(sign.to_string().as_bytes().to_vec())
+}
+
+pub fn verify<N: Network>(
+    address: &str,
+    signature: &Signature<N>,
+    message: &MessageDigest,
+) -> TofnResult<bool> {
+    use snarkvm::prelude::Address;
+
+    let message = aleo_encoded(message)?;
+
+    let address = Address::from_str(address).map_err(|_| TofnFatal)?;
+
+    Ok(signature.verify(&address, &message))
+}
+
+fn aleo_encoded<N: Network>(data: &MessageDigest) -> TofnResult<Vec<Field<N>>> {
+    let data = data.as_ref().try_into().map_err(|_| TofnFatal)?;
+    let num = cosmwasm_std::Uint256::from_le_bytes(data);
+    let message = format!("{num}group");
+
+    snarkvm::prelude::Value::from_str(message.as_str())
+        .map_err(|_| TofnFatal)?
```

Comment:
`error!` for consistency
https://github.com/eigerco/tofn/blob/dfaab09f9b9cc0bee52c4ddb15f0e893fd6fec27/src/ecdsa/mod.rs#L84-L86